### ISSUE TITLE
 Use manylinux2014 to get aarch64/ppc64le support

### ISSUE THREAD

**Issue by hrw** [scipy/scipy#11170]
Manylinux2014 image got released some time ago. One of things it brings is support for !x86 architectures.

Scipy is using Travis CI so two things can be joined and used to generate wheel files for aarch64 and ppc64le architectures.

Amount of software installed to be able to run 'pip install scipy' would be cut. Also install times would be nicer.

**Comment by rgommers** [comment]
There's limited tooling support and a bunch of TODOs on the rollout plan (https://github.com/pypa/manylinux/issues/338). So I'm wondering what the best plan is for producing wheels right now. It's too late for SciPy 1.4.0, which is ready to go. So we have 6 months till the next release.

Some questions:
- do we need to produce both `manylinux2010` and `manylinux2014` wheels for x86, or just one of them?
- do we want `ppcle64` wheels before fixing the knownfails and persistent crash we have at the moment?
- we'll need `aarch64` CI, do we use Drone or TravisCI or ...?

**Comment by tylerjereddy** [comment]
> we'll need aarch64 CI, do we use Drone or TravisCI or ...?

I think Drone has been working well for OpenBLAS ARM testing (@martin-frbg @isuruf ?). I feel like Shippable was causing quite a few issues for NumPy, though maybe that has quieted down?

**Comment by isuruf** [comment]
Drone has been working well for conda-forge. Few advantages over Travis-CI.
- Faster to boot
- 60 minutes per job instead of 50 mins
- Large limit on log size
- Travis has timeout of 10 minutes with no output, but drone doesn't have it (AFAIK)
- arm32 (armhf) support

**Comment by martin-frbg** [comment]
Yes, Drone has been working quite nicely for OpenBLAS since isuruf set it up.

**Comment by hrw** [comment]
@rgommers which manylinux to use is a question of which distros you support. Manylinux2014 is CentOS 7+ while manylinux2010 is CentOS 6+ (afaik).

**Comment by matthew-brett** [comment]
My suggestion would be to make manylinux2010 wheels until CentOS 6 goes
EOL, in November 30, 2020, and thereafter produce 2014 wheels.   Unless
there's some compelling gcc optimization we need before that.

**Comment by odidev** [comment]
Hi @rgommers as per the discussion https://github.com/scipy/scipy/issues/11724, I have modified the Travis.yml file to run builds for aarch64 ( currently only for python v3.7 ), but since the tests are taking longer than 50 minutes to execute the build is getting timed-out. Please have a look:

https://travis-ci.com/github/odidev/scipy/builds/157507705

I have already requested the Travis community to increase the timeout here: https://travis-ci.community/t/request-regarding-timeout-extension-for-arm64-builds/7938

Also, since you suggested that using drone-ci will be a much preferable option, I have been working on creating the drone.yml for aarch64 support ( for python v3.7 ), The tests have been successfully triggered but here also the test suite is taking longer than 60 minutes to complete hence build is halted: https://cloud.drone.io/odidev/scipy/13/1/2

I will be more than happy to contribute to making the aarch64 ci support available for scipy, Please share your thoughts on this.

**Comment by hrw** [comment]
Can tests be run in parallel? AArch64 machines usually have more cores than x86-64 ones.

**Comment by rgommers** [comment]
TravisCI gets to 65%, Drone only to 25% of the test suite - not what I had expected.

Tests can be run in parallel. If I look at what's happening now, `NPY_NUM_BUILD_JOBS=2` is set so that should speed up the build a little, however the tests aren't yet run in parallel. Installing `pytest-xdist` and running `python runtests.py --parallel=2` should fix that. I believe a TravisCI job gets 2 cores, although it would be worth double checking that for ARM64.

**Comment by isuruf** [comment]
Drone machines have 96 cores each, but is shared across projects.

**Comment by odidev** [comment]
@rgommers I am running the Travis-ci builds and facing the below issue while building it, could you please suggest what am I missing: https://travis-ci.com/github/odidev/scipy

**Comment by rgommers** [comment]
Probably removing `-u` from the runtests command, see https://stackoverflow.com/questions/14258500/python-significance-of-u-option. Best to keep modifications to a minimum I'd think, rather than throwing out so many build/test flags.

**Comment by odidev** [comment]
Thanks, @rgommers I have built scipy and here is the build log on Travis-ci for parallel=2 :
https://travis-ci.com/github/odidev/scipy/builds/159274043.

Since the above build was halting I have checked the same for parallel=3, Please have a look at the log: https://travis-ci.com/github/odidev/scipy

Also, I have tried the same on drone ci as well for parallel=3, Please have a look at the log:
https://cloud.drone.io/odidev/scipy/19/1/2

Please share your thoughts on this.

**Comment by rgommers** [comment]
Okay, so `parallel=2` still times out (just, gets to 98%) and `parallel=3` passes all tests except for:
```
$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ./tools/check_pyext_symbol_hiding.sh build; fi

build/testenv/lib/python3.7/site-packages/scipy/cluster/_vq.cpython-37m-aarch64-linux-gnu.so: too many public symbols!
0000000000005be4 T PyInit__vq
0000000000026718 B __pyx_module_is_main_scipy__cluster___vq
00000000000128d4 T _fini
0000000000002aa8 T _init
The command "if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ./tools/check_pyext_symbol_hiding.sh build; fi" exited with 1.
```

See gh-8463 for context on this test. The actual build log is hidden so it's hard to be completely sure, but it looks like it's using the conda-forge compilers and linker (`gcc-7.4` is also installed with `apt-get`, but conda compilers will be the first ones found normally). Those haven't been tested before on `aarch64`, so likely a real issue. Not clear to me where, could be in conda-forge's `binutils`, in `cython` or in `cluster.vq`.

@pv or @isuruf any thoughts?

**Comment by isuruf** [comment]
https://github.com/scipy/scipy/pull/11833

**Comment by odidev** [comment]
@rgommers I have been working on the Travis build for arm64, Please find the outcome for the same:

- The builds are getting successfully passed for arm64 platform with ```--parallel=3```, Please have a look: https://travis-ci.com/github/ossdev07/scipy/jobs/318718768.

**Note:-** when I have only triggered the two arm64 builds.

-  When running all the builds, all are getting successfully passed except the two arm64 related builds they are getting timed-out . I have tried increasing the ```--parallel=4``` and ```--parallel=5``` for arm64 but still getting timeout issue. Please have a look:
https://travis-ci.com/github/ossdev07/scipy

Please suggest what should possibly the next step here.

**Comment by isuruf** [comment]
@odidev, can you install `pytest-xdist` on drone as well and see if that helps?

**Comment by isuruf** [comment]
Nvm, it doesn't.

**Comment by rgommers** [comment]
>     * The builds are getting successfully passed for arm64 platform with `--parallel=3`, Please have a look: https://travis-ci.com/github/ossdev07/scipy/jobs/318718768.

Green in 48 minutes, nice.

> When running all the builds, all are getting successfully passed except the two arm64 related builds they are getting timed-out

Do we need two jobs? I'd be happy to have just the first one, and leave out `TESTMODE=full`. That may speed things up for the other job.

**Comment by odidev** [comment]
@rgommers I have raised as per your suggestion, Please have a look: https://github.com/scipy/scipy/pull/11867

**Comment by rgommers** [comment]
Okay, step 1 is done, we have ARM64 on TravisCI. Step 2 will be to add support for producing `manylinux2014` wheels to https://github.com/MacPython/scipy-wheels, and then we need to put them on PyPI for 1.5.0

**Comment by andyfaff** [comment]
Is there a reason we dont make wheel artefacts in the ci processes? It's good to have a defined process for releases, but it might be nice to make bleeding edge wheel artefacts available.

**Comment by charris** [comment]
ISTR that there were still some bugs in manylinux2014, at least for aarch64. @mattip Thoughts?

**Comment by andyfaff** [comment]
This is more of a general question for all OS and arches.

**Comment by rgommers** [comment]
Good point @andyfaff, once the migration from Rackspace to Anaconda is done, we should copy what NumPy does to provide latest wheels for CI purposes (see `tools/travis-upload-wheel.sh` in the numpy repo).

**Comment by mattip** [comment]
The gcc compiler was recently updated to one with a correct `copysign` implementation on aarch64, so at least compilation will run to completion.

There is still the problem that numpy blacklists certain glibc functions for the manylinux2014 glibc version (2.17). That means it replaces the problematic glibc implementations with local, numpy alternatives. On aarch64 the alternative `np.arcsinh` implementation fails some tests, and on ppc64le the np.float128 implementations use long double while ppc64 uses doubledouble, also leading to incorrect results for some values.

**Comment by rgommers** [comment]
Looks like for SciPy we're all good on `aarch64`; we do have a bunch of `ppc64le` `skipif`'s floating around that need fixing before we can provide wheels for that platform.

**Comment by mattip** [comment]
I am not sure SciPy aarch64 wheels on PyPI make sense without having all the dependency wheels (like NumPy) also available.

**Comment by odidev** [comment]
Ping @rgommers , I have created sample scipy aarch64 wheel and uploaded same on PyPi. Please find below supporting logs.

Travis Log:   https://travis-ci.com/github/mpmkp2020/scipy/jobs/327286542
travis file:  https://github.com/mpmkp2020/scipy/blob/aarch64-wheel/.travis.yml
cmake-odidev: https://pypi.org/project/scipy-mpmkp/

Could you please let me know if I can help in uploading scipy aarch64 wheel on PyPi?

**Comment by rgommers** [comment]
Thanks @odidev, looks good. The next step would be to integrate that with https://github.com/MacPython/scipy-wheels. That's the way we build wheels for all platforms every release, so we're sure it's a reproducible process. Could you try and add your changes there?

> There is still the problem that numpy blacklists certain glibc functions for the manylinux2014 glibc version (2.17).

I don't think that's a showstopper for SciPy, it would be good to get wheels up for the next release - those do have value even if there aren't any NumPy wheels yet.

**Comment by odidev** [comment]
Thanks @rgommers for quick response. I will work on integration with https://github.com/MacPython/scipy-wheels. Thanks.

**Comment by odidev** [comment]
Hi @rgommers, I am facing travis timeout issue. It looks like build/test scipy wheel on aarch64 is taking more than 50 minutes causing the aarch64 job exceeded the maximum time limit and gets aborted. Can you please help me by increasing maximum time limit for scipy-wheel for aarch64 job.

Travis-ci job run log: https://travis-ci.com/github/mpmkp2020/scipy-wheels/jobs/339344267
Travis file: https://github.com/mpmkp2020/scipy-wheels/blob/master/.travis.yml

**Comment by rgommers** [comment]
We do get timeouts in the CI of this repo as well.  TravisCI seems to have capacity issues unfortunately. We've already sped up the test suite a little, but it's not enough.

> Can you please help me by increasing the maximum time limit for scipy-wheel for aarch64 job.

I'm not sure we can do that on the free plan. Worth looking into, but I suspect it requires talking to TravisCI directly and them being quite generous.

 
> Travis-ci job run log: https://travis-ci.com/github/mpmkp2020/scipy-wheels/jobs/339344267

A problem here is that it's building a wheel for NumPy:
```
Building wheels for collected packages: numpy
```
That takes quite a few minutes. Maybe when re-running that same job it'll be cached. If not, we should put that wheel up on the Anaconda wheel store and grab it from there.

That may still not be enough, but it's worth doing. Another thing we could do here is separate the build and test parts into separate phases, but that's more work.

**Comment by hrw** [comment]
You can control what goes into cache on Travis. Add pip cache into it, do build of numpy wheel, cache it, next build should install it directly from cache.

**Comment by andyfaff** [comment]
NumPy wheels for both pypy and aarch64 are now available in the nightly builds.

**Comment by rgommers** [comment]
> NumPy wheels for both pypy and aarch64 are now available in the nightly builds.

That doesn't help us, we need lowest supported NumPy (1.14.5 right now) for building binaries in `scipy-wheels`.

**Comment by hrw** [comment]
1. Cut current job to just build numpy 1.14.5 wheel. 
2. Run such job and cache numpy wheel
3. enable build of scipy
4. run a scipy build - numpy wheel will came from cache

**Comment by odidev** [comment]
@hrw @rgommers @andyfaff I used numpy==1.14.5 aarch64 wheel present at https://files.pythonhosted.org/packages/0e/7a/10d4e79e0d141522736f41875e167041230c357351a512d2d8dcaeeb615d/numpy_mkp2020-1.14.5-cp37-cp37m-manylinux2014_aarch64.whl to save numpy build time. Even then travis-ci job is getting timeout. 

Travis-ci job run log: https://travis-ci.com/github/mpmkp2020/scipy-wheels/jobs/339891050

**Comment by mattip** [comment]
It seems both [shippable](https://www.shippable.com/github.html), which seem to use [Packet hardware](https://blog.shippable.com/shippable-arm-packet-deliver-native-ci-cd-for-arm-architecture) and [travis](https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system) run on [Packet hardware](https://www.packet.com) so probably not much to be gained by changing providers. In any case, even if we get it working, perhaps this CI job should not run on every PR since it could add a long delay.

**Comment by rgommers** [comment]
@mattip I'd expect it to be similar to `ppc64le` - initially flaky, but improves as TravisCI irons out bugs. If it gets too annoying we may have to temporarily disable it again, but I'd like to stick with it a little longer.

**Comment by rgommers** [comment]
@odidev it's not even finishing the build, which is weird. The build on this PR always completes in 15 min or so (or max 30 min). Maybe it's failing to build in parallel for you?

**Comment by mattip** [comment]
Wheel builds are taking [about 30 minutes average](https://travis-ci.org/github/MacPython/scipy-wheels) for X86, they end up calling `pip wheel` which may not be the optimal way to build: I don't see any flags for parallelism, and by default it uses an isolated PEP 517 build (which re-installs all the dependencies in an isolated environment).

**Comment by odidev** [comment]
@hrw @rgommers @andyfaff @mattip I tried using cache feature to speedup the build but even that is not benefiting us. Is there any procedure/flag to speedup the scipy build something like "make -j*".
Travis file: https://github.com/mpmkp2020/scipy-wheels/blob/master/.travis.yml

**Comment by rgommers** [comment]
The whole `multibuild` setup is quite convoluted, but if you follow all the commands it ends up at `build_wheel_cmd` in https://github.com/matthew-brett/multibuild/blob/a72a17b7a6150c74c927539564ab5badc4034321/common_utils.sh#L249. 

Unfortunately using `pip` rather that `python setup.py bdist_wheel` to build wheels makes it really difficult to pass in any build flags. The one suggestion I found is `.pydistutils.cfg` [here](https://stackoverflow.com/questions/15725869/pass-build-ext-options-to-pip-install), but I've never seen that before and am unsure if it works.

`multibuild` also goes out of its way to avoid `pyproject.toml`, so putting anything there won't help. Adding a `setup.cfg` next to scipy's `setup.py` within the build could work perhaps. 

Another thing that helps and isn't yet done is setting `NPY_NUM_BUILD_JOBS=2`. This is the easiest thing to try, because it's just an env var and will definitely be picked up.

**Comment by rgommers** [comment]
Hmm, in a quick test that doesn't actually go any faster:
```
$ git clean -xdf
$ export NPY_NUM_BUILD_JOBS=4
$ python setup.py build_ext -i
real    8m59.454s
user    13m11.372s
sys     0m31.967s

$ git clean -xdf
$ export NPY_NUM_BUILD_JOBS=2
$ python setup.py build_ext -i
real    9m50.316s
user    13m5.708s
sys     0m32.581s

$ git clean -xdf
$ unset NPY_NUM_BUILD_JOBS
$ python setup.py build_ext -i
real    8m36.636s
user    13m3.649s
sys     0m31.354s
```

The `-j2` flag will make it faster, but no good way to pass it in.

**Comment by mattip** [comment]
If you add a `config.sh` file you can override the [`pip_wheel_cmd`](https://github.com/matthew-brett/multibuild/blob/88a0b6f0eb770cf9f95792d66410e7696ce3d384/common_utils.sh#L241) function. Just be sure to copy the wheel to `$1` at the end, so something like
```
function pip_wheel_cmd {
    local abs_wheelhouse=$1
    # Assume hyperthreading?
    local CPUS=$(python -c "import multiprocessing; print(multiprocessing.cpu_count() / 2)")
    $PYTHON_EXE setup.py --build-only --parallel $CPUS
    cp dist/*.whl $abs_wheelhouse
}
```

**Comment by mattip** [comment]
With all that, it would be good to know what is taking all that time. Maybe we should start by adding a timestamp to the build log messages (in scipy/scipy, not here)

**Comment by odidev** [comment]
OK. I will try and update you on this.

**Comment by odidev** [comment]
@pv thanks for your suggestion, with the help of ccache I am able to build scipy wheel before travis-ci job gets timed out. I don't find any documentation for using f90cache in travis. Any suggestion on using both ccache/f90cache in travis-ci will be a great help?

@hrw, @rgommers, @andyfaff, @mattip, @pv now I am able to build AArch64 scipy wheel using "scipy-wheel". However, travis-ci job is getting timed out when executing "install_run $PLAT". Can you please help me to divide this job into two independent and serialized job to overcome this timeout issue?

**Comment by mattip** [comment]
It seems you would need to save the artifact somewhere. https://docs.travis-ci.com/user/build-stages/#data-persistence-between-stages-and-jobs

**Comment by mattip** [comment]
Maybe we should just build the wheel outside of CI and upload it to anaconda. @Qiyu8 offered use of an aarch64 machine.

**Comment by Qiyu8** [comment]
AFAIK,  The AArch64 wheel build process is timeout on travis-ci, I'm thinking of change another provider, there are two candidates suggested from the mailing list:

- [drone](https://drone.io/)
- [appveyor](https://www.appveyor.com/docs/server/#linux)

@odidev you have many experience on integrate with these continuous Delivery platforms, which one is better on AArch64 wheel build task? 
PS: I already own an aarch64 machine and ready to donate.

**Comment by andyfaff** [comment]
I've been [having a go](https://github.com/andyfaff/scipy/pull/16) using `drone`, and my experience is mixed. Each time you push a commit to a PR it does [two runs](https://cloud.drone.io/andyfaff/scipy). I think the first is the PR, and the second is the PR as it would be merged into master. The former timed out (60 mins), and the second was fine (~40mins). I tried rerunning the former twice and it timed out on both occasions. The availability of a numpy wheel would speed things up by a few (~6?) minutes.

**Comment by andyfaff** [comment]
conda-forge is now using drone to build arm64 packages, perhaps their experience would be illuminating.

**Comment by isuruf** [comment]
Drone has 192 small core machines and 88 large core machines. If it's the large core machine, you don't get time outs.

**Comment by andyfaff** [comment]
@isuruf is it possible to request the latter?

**Comment by odidev** [comment]
@pv, @rgommers shall I build aarch64 wheel and will you upload same on pypi or shall I use another build platform like drone or appveyor. Can you please let me know your thought on this?

**Comment by pv** [comment]
The build process should be made reproducible so that the release manager can run it (and so that it is available for future versions). I'd prefer to not upload things to pypi that do not come from our build infrastructure / RM. The preferable I think is to make it a part of scipy-wheels.

**Comment by odidev** [comment]
Thanks @pv for a quick response, I will work on your suggestions and get back to you.

**Comment by odidev** [comment]
@pv, I modified the scipy-wheel travis file to divide the task in tow stages to build and test separately. building is done successful but the test is failing with below issue after completing upto 99% cases .

'No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.'

Please have a look at below logs/travis file.
travis.yml: [travis.yml](https://github.com/odidev/scipy-wheels/blob/master/.travis.yml)
logs: [logs](https://travis-ci.com/github/odidev/scipy-wheels/builds/174190024 )

Can you please help me to resolve this issue?

**Comment by rgommers** [comment]
@odidev it looks to me like there's a test that's run at 99% of the test suite that is extremely slow for some reason. Could you add `verbose=2` to the test command at https://github.com/odidev/scipy-wheels/blob/master/run_scipy_tests.py#L19? That will show the test name it hangs on.

**Comment by AGSaidi** [comment]
I noticed that some of the OpenBLAS regressions ran for an extremely long time and sent them some PRs that fix those issues which were released with OpenBLAS 0.3.9.  Maybe there are still some problems there.

**Comment by odidev** [comment]
@rgommers, it looks like Travis-ci build log is limited to just 4 MB. Adding verbose=2 is displaying all the test name and this results in crossing the size of log limit after completing around 65% of test cases. Is there any way to display just part of test name(ignoring classs name)?

**Comment by mattip** [comment]
Could you use a per-directory conftest.py (used by pytest) to increase verbosity only for the directory in question? Here is a place to start looking at this in the documentation https://docs.pytest.org/en/stable/logging.html

**Comment by rgommers** [comment]
Usually `pytest` runs tests in the same order, the 99% is almost certainly in `scipy.stats`. So you can modify the test run like so:
```
import scipy.stats

pytest_args = args.pytest_args
pytest_args += ['durations=20']
ret = scipy.stats.test(args.test_mode, extra_argv=pytest_args)
```

**Comment by odidev** [comment]
@rgommers, as per your suggestion, I modified the travis file and it looks like it successfully passed scipy.stats tests. However, it is failing while uploading the wheel. Please find below logs link-
Travis-logs: [logs](https://travis-ci.com/github/odidev/scipy-wheels/jobs/358811631)
Travis.yml: [travis.yml](https://github.com/odidev/scipy-wheels/blob/master/.travis.yml)
Test run script: [run_scipy_tests.py](https://github.com/odidev/scipy-wheels/blob/master/run_scipy_tests.py#L24)

Can you please suggest where shall I investigate further?

**Comment by mattip** [comment]
The upload failure is to be expected, you do not have a token, so the command is ill-formed. 

This test times out after 30 minutes, https://travis-ci.com/github/odidev/scipy-wheels/jobs/358712783#L186, perhaps there is a time limit somewhere? The one that reached 99% also seems to have taken 30 minutes. https://travis-ci.com/github/odidev/scipy-wheels/builds/174190024

**Comment by odidev** [comment]
[Here](https://travis-ci.com/github/odidev/scipy-wheels/jobs/358712783#L186) the test was hitting timeout issue(didn't received anything to display in last 10 minute). So, I set it to 30 minute.

**Comment by odidev** [comment]
@rgommers, any suggestion on the above [comment](https://github.com/scipy/scipy/issues/11170#issuecomment-655601641)

**Comment by rgommers** [comment]
So here are the slowest 20 tests from running the `stats` test suite successfully (from your log):
```
========================== slowest 20 test durations ===========================
111.67s call     tests/test_continuous_basic.py::test_cont_basic[kstwo-arg59]
95.25s call     tests/test_stats.py::TestMGCStat::test_twosamp
75.69s call     tests/test_distributions.py::TestLevyStable::test_pdf_nolan_samples
75.42s call     tests/test_continuous_basic.py::test_moments[vonmises-arg100-False-False-False]
65.77s call     tests/test_continuous_basic.py::test_moments[kstwo-arg59-True-True-False]
52.80s call     tests/test_continuous_basic.py::test_moments[vonmises_line-arg102-True-True-False]
40.61s call     tests/test_continuous_basic.py::test_moments[recipinvgauss-arg87-True-True-False]
36.90s call     tests/test_continuous_basic.py::test_moments[vonmises-arg101-True-True-True]
34.23s call     tests/test_continuous_basic.py::test_cont_basic[genexpon-arg30]
32.83s call     tests/test_continuous_basic.py::test_moments[kappa4-arg53-True-True-False]
29.86s call     tests/test_continuous_basic.py::test_moments[genexpon-arg30-True-True-False]
29.27s call     tests/test_continuous_basic.py::test_moments[kappa4-arg55-True-True-False]
26.75s call     tests/test_stats.py::TestMGCStat::test_oned[independence--0.0094-0.78]
26.70s call     tests/test_stats.py::TestMGCStat::test_fived[nonlinear-0.019-0.117]
26.45s call     tests/test_stats.py::TestMGCStat::test_fived[linear-0.184-0.001]
26.26s call     tests/test_stats.py::TestMGCStat::test_random_state
25.38s call     tests/test_continuous_basic.py::test_moments[kappa4-arg56-True-True-False]
25.29s call     tests/test_continuous_basic.py::test_moments[ksone-arg58-True-True-False]
24.23s call     tests/test_stats.py::TestMGCStat::test_oned[nonlinear-0.163-0.001]
23.67s call     tests/test_stats.py::TestMGCStat::test_oned[linear-0.97-0.001]
```

All of those tests are marked with `pytest.mark.slow`, so they shouldn't be run at all for `aarch64` here. For some reason `args.test_mode` ([here](https://github.com/odidev/scipy-wheels/blob/master/run_scipy_tests.py#L21)) in your CI run is `'full'` rather than the default of `'fast'`. If you track that down and change it, the tests should fit within the timeout limit easily.

**Comment by odidev** [comment]
@rgommers, for now I have hard code test_mode to 'fast' to see if all the test cases are passing. I observed that there are 4 test cases which are failing. Please have a look at [this](https://travis-ci.com/github/odidev/scipy-wheels/jobs/361689956). Below are the failed test cases-

FAILED optimize/tests/test_linprog.py::TestLinprogSimplexNoPresolve::test_bounds_infeasible_2
FAILED linalg/tests/test_lapack.py::test_orcsd_uncsd[complex128] - AssertionE...
FAILED linalg/tests/test_build.py::TestF77Mismatch::test_lapack - AssertionEr...
FAILED optimize/tests/test__remove_redundancy.py::TestRRSVD::test_magic_square2

Can you please help me resolving this issue?

**Comment by rhenwood-arm** [comment]
(I work with @odidev ) @odidev please can you help me understand where and what you think the problem is with the AArch64 build. I have been following this thread but it isn't clear to me what problem you're trying to solve current.

**Comment by mattip** [comment]
I would start with this one:
```
/venv/lib/python3.7/site-packages/scipy/linalg/tests/test_build.py:55: in test_lapack
E   AssertionError: Both g77 and gfortran runtimes linked in scipy.linalg.flapack ! This is
E   likely to cause random crashes and wrong results. See numpy INSTALL.rst.txt for
E   more information.
        deps       = [b'libgfortran', b'libgfortran']
        f          = <scipy.linalg.tests.test_build.FindDependenciesLdd object at 0xffff5c913810>
        self       = <scipy.linalg.tests.test_build.TestF77Mismatch object at 0xffff5c913650>
```

**Comment by odidev** [comment]
@rhenwood-arm, I will look into the problem and share my findings with you. 

Let me explain you the current work...
Recently, while running the test cases, I was facing travis timeout due to no response received in last 10 minutes. @rgommers helped me to identify and skip such test cases. I was doing this exercise and observed few tests failure. Now, I am looking into these tests failure.

**Comment by hrw** [comment]
According to https://github.com/pyca/cryptography/issues/5292#issuecomment-667349697 you may try new aarch64 nodes on travis ci (if you are on .com not .org). Those are AWS Graviton2 based so should be faster.

**Comment by odidev** [comment]
@rgommers, almost all the test cases are passing on AArch64 machine except one and the same test is failing on  x86_64 machine as well. 

**Below are the steps to resolve aarch64 issue:**
1) As suggested by @rgommers, ran the tests in fast test_mode instead of full test_mode.
2) Installed https://anaconda.org/multibuild-wheels-staging/openblas-libs/v0.3.9/download/openblas-v0.3.9-manylinux2014_aarch64.tar.gz openblas via inbuilt function instead of installing threw apt/yum. This needed to add SHA-256 value of openblas-v0.3.9-manylinux2014_aarch64.tar.gz file in sha256_vals list at line [openblas_support.py#L37](https://github.com/odidev/scipy/blob/master/tools/openblas_support.py#L37)


**Below is the failing test case on both the platform**
**failure log snippet-**
_/venv/lib/python3.6/site-packages/scipy/optimize/_linprog_simplex.py:166: in _pivot_row
    return True, min_rows[0]
E   IndexError: index 0 is out of bounds for axis 0 with size 0
        T          = array([[ 1.,  0.,  0.,  1.,  0.,  1.,  0., nan],
       [ 0.,  0.,  1.,  0.,  1.,  0.,  1., nan],
       [ 1.,  2.,  3.,  0.,  0.,  0.,  0., inf],
       [-1., -0., -1., -1., -1.,  0.,  0., nan]])_

Failure on x86_64:
    https://travis-ci.com/github/odidev/scipy-wheels/builds/179065731
	
Failure on AArch64:
    https://travis-ci.com/github/odidev/scipy-wheels/jobs/369550585
	
**Can you please let me know if we need to solve this issue or else this is know and accepted? I am unable to verify this as I can't see any build run log link on scipy-wheel repository.**

Thanks

**Comment by rgommers** [comment]
Interesting, that test failure doesn't seem like a blocker for ARM64 wheels then, but it would be good to figure it out (possibly in a separate issue). Did you change the OpenBLAS version for `x86_64`? If not, then I'm not sure why we don't see it on master.

It seems like you're at the point where we can merge your work @odidev, very nice. Can you open a PR with your changes? Please leave that one failure as is, and open a separate issue in this repo for it. Then we'll figure it out there; we may add an `xfail` to it temporarily to make the wheel build pass.

**Comment by odidev** [comment]
@rgommers, no, I did not changed OpenBLAS version. I just used AArch64 specific version of OpenBLAS in scipy. To do this exercise, I forked scipy and added AArch64 specific version of OpenBLAS into it and integrated this forked scipy as git submodule in scipy-wheel repository. 

Here, first we need to open a PR in scipy and then in scipy-wheel. I am working on it. Thanks.

**Comment by odidev** [comment]
@tylerjereddy, @rgommers It looks like this issue got auto closed due to merging [PR-12705](https://github.com/scipy/scipy/pull/12705.) Can you please re-open this issue?

**Comment by tylerjereddy** [comment]
The ARM64 wheels PR is getting a bit closer I think: https://github.com/MacPython/scipy-wheels/pull/93

I've added some minor comments there just now. A few other thoughts:
- do we plan to try to backport to 1.5.x to release ARM64 wheels before 1.6.0?
- travis ci .com instead of the older travisci.org apparently allows for better/faster ARM nodes, but can we reasonably expect all of MacPython to migrate over soon-ish?

**Comment by rgommers** [comment]
> do we plan to try to backport to 1.5.x to release ARM64 wheels before 1.6.0?

This would be helpful for quite a few users, but I also have a memory of breaking people's deployment pipelines when we added wheels for old releases once for NumPy. I can't remember why though. I'd be inclined to not do this, IIRC we once decided it was a bad idea to go amend old releases.

**Comment by matthew-brett** [comment]
As I remember it, when I uploaded Intell-architecture Numpy wheels for previous releases, it found someone whose pipeline depended on the fact that Numpy was installing from source for an older release.   I suspect this is very unusual.

**Comment by tylerjereddy** [comment]
To clarify, if we release a new `1.5.3` point release, it could still break the workflow for people because it has one more binary/wheel available vs. `1.5.2`?

I'm not suggesting we add an extra binary/wheel to something that has already been released, but just have another point release that adds ARM64 wheels alongside new wheels of the usual type.

**Comment by tylerjereddy** [comment]
See also some discussion here: https://github.com/MacPython/scipy-wheels/pull/93#issuecomment-700469843 

We should also keep in mind that while I should probably do a 1.5.3 release because of XCode 12 compatibility break anyway, backporting ARM64 shims/fixes on both the main repo and wheels repo is sure to run into some hiccups.

**Comment by tylerjereddy** [comment]
One might even wonder if adding support for a new architecture is considered a "feature" that falls outside the scope of a point release semantic versioning-like approach, though that's perhaps a bit formal. The main concern I have is just quite experienced folks above indicating that problems have happened in the past.

**Comment by rgommers** [comment]
I think adding the wheel for `1.5.3` makes sense. If it breaks, we fix it or we pull it quickly.

**Comment by AGSaidi** [comment]
Personally, I'd love to see an 1.5.3 arm64 wheel too. @tylerjereddy if the time to backport is the concern, I think you'll have some volunteers to help.

**Comment by tylerjereddy** [comment]
Okie, we're working on it

**Comment by mattkanwisher** [comment]
Seems like aarch64 is only for Macs? There is a lot of machine learning happening on the edge with SciPy, it currently takes over an hour to install on embedded arm devices. Any chance we can get wheels for linux aarch64 also?

**Comment by mattip** [comment]
> Seems like aarch64 is only for Macs?

The wheels are called manylinux2014 because they are for linux. MacOS11 hybrid (or whatever they call it) is a different kettle of fish to fry.

**Comment by rgommers** [comment]
@tylerjereddy `manylinux2014` wheels for `aarch64` are up, and `ppc64le` is probably not worth doing at this point given new TravisCI plan limitations. So this can be closed?

**Comment by tylerjereddy** [comment]
Alright, I'll close it then. The time/energy needed to deal with that Travis CI shift and also expand (or even just maintain) our wheels offerings is maybe something worthy of a mention in a maintenance grant.

**Comment by rgommers** [comment]
> The time/energy needed to deal with that Travis CI shift and also expand (or even just maintain) our wheels offerings is maybe something worthy of a mention in a maintenance grant.

Yes I agree. It's one of the most essential as well as most painful things that needs doing in SciPy.
